aws-api/src/main/java/com/amplifyframework/api/aws/AWSApiPlugin.java

```diff
@@ -732,4 +748,43 @@ public int hashCode() {
             return result;
         }
     }
+
+    /**
+     * This class implements OkHttp's {@link EventListener}. Its main purpose
+     * is to listen to network-related events reported by the http client and trigger
+     * a Hub event if necessary.
+     */
+    private final class ApiConnectionEventListener extends EventListener {
+        private AtomicReference<ApiEndpointStatus> currentNetworkStatus;
+
+        ApiConnectionEventListener() {
+            currentNetworkStatus = new AtomicReference<>(ApiEndpointStatus.UNKOWN);
+        }
+
+        @Override
+        public void connectFailed(@NotNull Call call,
+                                  @NotNull InetSocketAddress inetSocketAddress,
+                                  @NotNull Proxy proxy,
+                                  @Nullable Protocol protocol,
+                                  @NotNull IOException ioe) {
+            super.connectFailed(call, inetSocketAddress, proxy, protocol, ioe);
+            triggerEventIfStatusChanged(currentNetworkStatus.getAndSet(ApiEndpointStatus.NOT_REACHABLE));
+        }
+
+        @Override
+        public void connectionAcquired(@NotNull Call call, @NotNull Connection connection) {
+            super.connectionAcquired(call, connection);
+            triggerEventIfStatusChanged(currentNetworkStatus.getAndSet(ApiEndpointStatus.REACHABLE));
+        }
+
+        private void triggerEventIfStatusChanged(@NonNull ApiEndpointStatus previousStatus) {
+            if (!previousStatus.equals(currentNetworkStatus.get())) {
+                ApiEndpointStatusChangeEvent eventData =
+                    new ApiEndpointStatusChangeEvent(currentNetworkStatus.get(), previousStatus);
```

Comment:
You call `currentNetworkStatus.get()` twice here, without any synchronization guards, it appears. What happens if the value changes between the two calls?

1. Do you need to add syncrhonization primitives?
2. You can defend by making one `.get()`, caching the result, and then using the cached result in both locations.

Reply:
Good call. I'm going to implement the latter.